### PR TITLE
Fix four bugs with numeric dimension output types.

### LINKED ISSUE #6123: ClassCastException in groupBy when sorting on numeric columns containing nulls
Here is the sample query.

```
{
  "queryType": "groupBy",
  "dataSource": {
    "type": "table",
    "name": "wikiticker"
  },
  "intervals": {
    "type": "intervals",
    "intervals": [
      "-146136543-09-08T08:23:32.096Z/146140482-04-24T15:36:27.903Z"
    ]
  },
  "virtualColumns": [
    {
      "type": "expression",
      "name": "d0:v",
      "expression": "timestamp_floor(\"__time\",'PT1H','','UTC')",
      "outputType": "LONG"
    }
  ],
  "filter": null,
  "granularity": {
    "type": "all"
  },
  "dimensions": [
    {
      "type": "default",
      "dimension": "d0:v",
      "outputName": "d0",
      "outputType": "LONG"
    },
    {
      "type": "default",
      "dimension": "cityName",
      "outputName": "d1",
      "outputType": "STRING"
    }
  ],
  "aggregations": [],
  "postAggregations": [],
  "having": null,
  "limitSpec": {
    "type": "default",
    "columns": [
      {
        "dimension": "d0",
        "direction": "descending",
        "dimensionOrder": {
          "type": "numeric"
        }
      },
      {
        "dimension": "d1",
        "direction": "ascending",
        "dimensionOrder": {
          "type": "lexicographic"
        }
      }
    ],
    "limit": 5000
  },
  "context": {
    "timeout": 300000
  },
  "descending": false
}
```

If `__time` column contains nulls, this query throws an error like below.

```
java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.Long
        at java.lang.Long.compareTo(Long.java:54) ~[?:1.8.0_163]
        at com.google.common.collect.NaturalOrdering.compare(NaturalOrdering.java:35) ~[guava-16.0.1.jar:?]
        at com.google.common.collect.NaturalOrdering.compare(NaturalOrdering.java:26) ~[guava-16.0.1.jar:?]
        at com.google.common.collect.NullsFirstOrdering.compare(NullsFirstOrdering.java:44) ~[guava-16.0.1.jar:?]
        at io.druid.query.groupby.GroupByQuery.compareDimsForLimitPushDown(GroupByQuery.java:582) ~[druid-processing-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.query.groupby.GroupByQuery.access$000(GroupByQuery.java:81) ~[druid-processing-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.query.groupby.GroupByQuery$1.compare(GroupByQuery.java:416) ~[druid-processing-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.query.groupby.GroupByQuery$1.compare(GroupByQuery.java:412) ~[druid-processing-0.12.1-iap8.jar:0.12.1-iap8]
        at com.google.common.collect.ComparatorOrdering.compare(ComparatorOrdering.java:38) ~[guava-16.0.1.jar:?]
        at io.druid.common.guava.CombiningSequence$CombiningAccumulator.accumulate(CombiningSequence.java:260) ~[druid-common-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.YieldingAccumulators$1.accumulate(YieldingAccumulators.java:35) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.MergeSequence.makeYielder(MergeSequence.java:106) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.MergeSequence.toYielder(MergeSequence.java:94) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.LazySequence.toYielder(LazySequence.java:46) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.query.RetryQueryRunner$1.toYielder(RetryQueryRunner.java:102) ~[druid-processing-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.YieldingSequenceBase.accumulate(YieldingSequenceBase.java:32) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.common.guava.CombiningSequence.accumulate(CombiningSequence.java:64) ~[druid-common-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.MappedSequence.accumulate(MappedSequence.java:43) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.query.groupby.orderby.TopNSequence$1.make(TopNSequence.java:55) ~[druid-processing-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.BaseSequence.toYielder(BaseSequence.java:65) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.WrappingSequence$2.get(WrappingSequence.java:87) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.WrappingSequence$2.get(WrappingSequence.java:83) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.SequenceWrapper.wrap(SequenceWrapper.java:55) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.WrappingSequence.toYielder(WrappingSequence.java:82) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.MappedSequence.toYielder(MappedSequence.java:49) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.WrappingSequence$2.get(WrappingSequence.java:87) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.WrappingSequence$2.get(WrappingSequence.java:83) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.query.CPUTimeMetricQueryRunner$1.wrap(CPUTimeMetricQueryRunner.java:74) ~[druid-processing-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.WrappingSequence.toYielder(WrappingSequence.java:82) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.java.util.common.guava.Yielders.each(Yielders.java:32) ~[java-util-0.12.1-iap8.jar:0.12.1-iap8]
        at io.druid.server.QueryResource.doPost(QueryResource.java:193) [druid-server-0.12.1-iap8.jar:0.12.1-iap8]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_163]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_163]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_163]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_163]2018-08-07T20:49:53,620
```

[druid-0.12.1-iap8](https://github.com/implydata/druid/tree/druid-0.12.1-iap8) is almost same with 0.12.2.

### PR DESCRIPTION
**Pull request by gianm**
This patch includes the following bug fixes:

- TopNColumnSelectorStrategyFactory: Cast dimension values to the output type
  during dimExtractionScanAndAggregate instead of updateDimExtractionResults.
  This fixes a bug where, for example, grouping on doubles-cast-to-longs would
  fail to merge two doubles that should have been combined into the same long value.
- TopNQueryEngine: Use DimExtractionTopNAlgorithm when treating string columns
  as numeric dimensions. This fixes a similar bug: grouping on string-cast-to-long
  would fail to merge two strings that should have been combined.
- GroupByQuery: Cast numeric types to the expected output type before comparing them
  in compareDimsForLimitPushDown. This fixes #6123.
- GroupByQueryQueryToolChest: Convert Jackson-deserialized dimension values into
  the proper output type. This fixes an inconsistency between results that came
  from cache vs. not-cache: for example, Jackson sometimes deserializes integers
  as Integers and sometimes as Longs.

And the following code-cleanup changes, related to the fixes above:

- DimensionHandlerUtils: Introduce convertObjectToType, compareObjectsAsType,
  and converterFromTypeToType to make it easier to handle casting operations.
- TopN in general: Rename various "dimName" variables to "dimValue" where they
  actually represent dimension values. The old names were confusing.

### CODE REVIEW

File: processing/src/main/java/io/druid/query/topn/PooledTopNAlgorithm.java [hunk 1]
```diff
@@ -749,14 +743,9 @@ protected void updateResults(
           position += aggregatorSizes[j];
         }
 
-        Object retVal = dimSelector.lookupName(i);
-        if (needsResultConversion) {
-          retVal = valueTransformer.apply(retVal);
-        }
-
-
+        // Output type must be STRING in order for PooledTopNAlgorithm to make sense; so no need to convert value.
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/topn/PooledTopNAlgorithm.java]: Maybe add an assert about the output type here.

File: processing/src/main/java/io/druid/query/topn/types/StringTopNColumnSelectorStrategy.java [hunk 2]
```diff
@@ -19,34 +19,36 @@
...
 import java.util.Map;
+import java.util.function.Function;
 
 public class StringTopNColumnSelectorStrategy
-    implements TopNColumnSelectorStrategy<DimensionSelector, Map<String, Aggregator[]>>
+    implements TopNColumnSelectorStrategy<DimensionSelector, Map<Comparable, Aggregator[]>>
 {
-  @Override
-  public int getCardinality(DimensionSelector selector)
+  private final Function<Object, Comparable<?>> dimensionValueConverter;
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/topn/types/StringTopNColumnSelectorStrategy.java]: Function from String?

File: processing/src/main/java/io/druid/query/topn/types/StringTopNColumnSelectorStrategy.java [hunk 3]
```diff
@@ -126,7 +119,7 @@ private long dimExtractionScanAndAggregateWithCardinalityKnown(
       Cursor cursor,
       DimensionSelector selector,
       Aggregator[][] rowSelector,
-      Map<String, Aggregator[]> aggregatesStore
+      Map<Comparable, Aggregator[]> aggregatesStore
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/topn/types/StringTopNColumnSelectorStrategy.java]: `Comparable` and `Comparable<?>` are used inconsistently in this class

File: processing/src/main/java/io/druid/query/topn/types/TopNColumnSelectorStrategyFactory.java [hunk 4]
```diff
@@ -27,23 +28,39 @@
...
+        return new StringTopNColumnSelectorStrategy(dimensionType);
       case LONG:
-        return new NumericTopNColumnSelectorStrategy.OfLong();
       case FLOAT:
-        return new NumericTopNColumnSelectorStrategy.OfFloat();
       case DOUBLE:
-        return new NumericTopNColumnSelectorStrategy.OfDouble();
+        if (ValueType.isNumeric(dimensionType)) {
+          // Return strategy that aggregates using the _output_ type, because this allows us to collapse values
+          // properly (numeric types cannot represent all values of other numeric types).
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/topn/types/TopNColumnSelectorStrategyFactory.java]: FWIW `double` represents all values of `float`

File: processing/src/main/java/io/druid/query/topn/types/TopNColumnSelectorStrategyFactory.java [hunk 5]
```diff
@@ -27,23 +28,39 @@
...
+          return NumericTopNColumnSelectorStrategy.ofType(dimensionType, dimensionType);
+        } else {
+          // Return strategy that aggregates using the _input_ type. Here we are assuming that the output type can
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/topn/types/TopNColumnSelectorStrategyFactory.java]: Despite those comments, it's still hard to understand what is going on here. Maybe you could try to clarify it more?

File: processing/src/main/java/io/druid/segment/DimensionHandlerUtils.java [hunk 6]
```diff
@@ -302,6 +306,64 @@ public static Float convertObjectToFloat(@Nullable Object valObj, boolean report
...
+    //noinspection unchecked
+    return Comparators.<Comparable>naturalNullsFirst().compare(
+        convertObjectToType(lhs, type),
+        convertObjectToType(rhs, type)
+    );
+  }
+
+  @Nullable
+  public static Comparable<?> convertObjectToType(
+      @Nullable final Object obj,
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/DimensionHandlerUtils.java]: Unnecessary breakdown

**Comment by leventov** [comment]
@fjy please don't merge PRs with unanswered comments (despite the approval), otherwise why they are left?

**Comment by gianm** [comment]
Fwiw, I had already started addressing some of those comments, so I plan to raise a follow up PR for that.

Btw, @leventov what did you intend by doing an approval + also leaving comments? I'd interpret that as "please consider these comments, but if you don't want to do them, I am ok with that." I thought they were all reasonable comments so that's why I'm doing a follow up.

**Comment by gianm** [comment]
Follow-up is in #6231.

**Pull request merged**